login/login.go

```diff
@@ -0,0 +1,405 @@
+	FamilyName        string   `json:"family_name"`
+	Email             string   `json:"email"`
+	EmailVerified     bool     `json:"email_verified"`
+	Ga4ghPassportV1   []string `json:"ga4gh_passport_v1"`
+}
+
+type DeviceLogin struct {
+	BaseURL         string
+	ClientID        string
+	S3Target        string
+	PollingInterval int
+	LoginResult     *Result
+	UserInfo        *UserInfo
+	wellKnown       *OIDCWellKnown
+	deviceLogin     *DeviceLoginResponse
+}
+
+type AuthInfo struct {
+	ClientID  string `json:"client_id"`
+	OidcURI   string `json:"oidc_uri"`
+	PublicKey string `json:"public_key"`
+	InboxURI  string `json:"inbox_uri"`
+}
+
+// requests the /info endpoint to fetch the parameters needed for login
+func GetAuthInfo(baseURL string) (*AuthInfo, error) {
+	url := baseURL + "/info"
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	var result AuthInfo
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(body, &result)
+	if err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+// creates a .sda-cli-session file and updates its values
+func (login *DeviceLogin) UpdateConfigFile() error {
+
+	_, err := os.Create(".sda-cli-session")
+	if err != nil {
+		return err
+	}
+	cfg, err := ini.Load(".sda-cli-session")
+	if err != nil {
+		return err
+	}
+
+	s3Config, err := login.GetS3Config()
+	if err != nil {
+		return err
+	}
+
+	err = ini.ReflectFrom(cfg, s3Config)
+	if err != nil {
+		return err
+	}
+	err = cfg.SaveTo(".sda-cli-session")
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// NewDeviceLogin() returns a new `DeviceLogin` with the given `url` and
+// `clientID` set.
+func NewDeviceLogin(args []string) (DeviceLogin, error) {
+
+	var url string
+	err := Args.Parse(args[1:])
+	if err != nil {
+		return DeviceLogin{}, errors.New("failed parsing arguments")
+	}
+	if len(Args.Args()) == 1 {
+		url = Args.Args()[0]
+	}
+	info, err := GetAuthInfo(url)
+	if err != nil {
+		return DeviceLogin{}, errors.New("failed to get auth Info")
+	}
+
+	return DeviceLogin{BaseURL: info.OidcURI, ClientID: info.ClientID, PollingInterval: 2, S3Target: info.InboxURI}, nil
+}
+
+// open opens the specified URL in the default browser of the user.
+func open(url string) error {
+	var cmd string
+	var args []string
+
+	switch runtime.GOOS {
+	case "windows":
+		cmd = "cmd"
+		args = []string{"/c", "start"}
+	case "darwin":
+		cmd = "open"
+	default: // "linux", "freebsd", "openbsd", "netbsd"
+		cmd = "xdg-open"
+	}
+	args = append(args, url)
+
+	return exec.Command(cmd, args...).Start()
+}
+
+// Login() does a full login by fetching the remote configuration, starting the
+// login procedure, and then waiting for the user to complete login.
+func (login *DeviceLogin) Login() error {
+
+	var err error
+	login.wellKnown, err = login.getWellKnown()
+	if err != nil {
+		return fmt.Errorf("failed to fetch .well-known configuration: %v", err)
```

Comment:
It's better to show the message as 

```
failed to fetch .well-known configuration, details: %v", err
```
similar as https://github.com/NBISweden/sda-cli/blob/main/download/download.go#L79 and https://github.com/NBISweden/sda-cli/blob/main/download/download.go#L90

Also, it is probably better to add a line break at the end of message, so that multiple messages will not stack in one line. 

This applies to all error messages in this script.

Reply:
I am not sure about the newline:

> error strings should not end with punctuation or newlines (ST1005)

But I see your point of it becoming a bit long with the errors following.
It seems like this is how we usually return the errors so far, error string followed by a colon and the actual error.
For example: https://github.com/NBISweden/sda-cli/blob/227d4d291b0991250236a20ff9ab9afa0919ab79/create_key/create_key.go#L54

Reply:
It's not just the line becomes long but also messages are merged since there is no space in between. For example 

```
$ ./sda-cli login https://login.bp.nbis.se/
failed to contact authentication service: failed to get auth InfoLogin failed: failed to start device login: request failed: status code: 400
```
The message `failed to get auth Info` and `Login failed: failed to start device login: ` should be separated.

Reply:
Okay, I see now which ones are the problem. I'll try to improve it.